File: meta-platform/backend/app/schemas/meta.py
from typing import Any, Literal

from pydantic import BaseModel, Field


class FieldConstraints(BaseModel):
    # text / textarea
    min_length: int | None = Field(default=None, alias="minLength")
    max_length: int | None = Field(default=None, alias="maxLength")
    pattern: str | None = None

    # number / progress
    min: float | None = None
    max: float | None = None
    step: float | None = None

    # date
    min_date: str | None = Field(default=None, alias="minDate")
    max_date: str | None = Field(default=None, alias="maxDate")

    # multiselect / tags
    min_items: int | None = Field(default=None, alias="minItems")
    max_items: int | None = Field(default=None, alias="maxItems")

    model_config = {"populate_by_name": True}


class XUIHints(BaseModel):
    widget: str | None = None
    kanban_group: bool = False
    color_map: dict[str, str] = {}
    placeholder: str | None = None
    readonly: bool = False
    width: str | None = None


class FieldMeta(BaseModel):
    name: str
    type: Literal["text", "textarea", "select", "multiselect", "date", "user", "tags", "number", "progress", "checkbox"]
    label: str
    required: bool = False
    options: list[str] | None = None
    default: Any = None
    x_ui: XUIHints = Field(default_factory=XUIHints, alias="x-ui")
    constraints: FieldConstraints | None = None

    model_config = {"populate_by_name": True}


class ActionMeta(BaseModel):
    id: str
    label: str
    icon: str | None = None
    variant: Literal["primary", "secondary", "danger"] = "secondary"


class EntityMeta(BaseModel):
    id: str
    name: str
    description: str = ""
    layout: Literal["kanban", "table", "form", "grid"] = "kanban"
    fields: list[FieldMeta]
    actions: list[ActionMeta] = []
    allow_attachments: bool = False
    version: int = 1

File: meta-platform/frontend/src/types/meta.ts
export type FieldType =
  | 'text'
  | 'textarea'
  | 'select'
  | 'multiselect'
  | 'date'
  | 'user'
  | 'tags'
  | 'number'
  | 'progress'
  | 'checkbox'

export type LayoutType = 'kanban' | 'table' | 'form' | 'grid'
export type ActionVariant = 'primary' | 'secondary' | 'danger'

export interface XUIHints {
  widget?: string
  kanban_group?: boolean
  color_map?: Record<string, string>
  placeholder?: string
  readonly?: boolean
  width?: string
}

export interface FieldConstraints {
  // text / textarea
  minLength?: number
  maxLength?: number
  pattern?:   string   // regex

  // number / progress
  min?:  number
  max?:  number
  step?: number

  // date
  minDate?: string   // YYYY-MM-DD
  maxDate?: string   // YYYY-MM-DD

  // multiselect / tags
  minItems?: number
  maxItems?: number
}

export interface FieldMeta {
  name: string
  type: FieldType
  label: string
  required?: boolean
  options?: string[]
  default?: unknown
  'x-ui': XUIHints
  constraints?: FieldConstraints
}

export interface ActionMeta {
  id: string
  label: string
  icon?: string
  variant: ActionVariant
}

export interface EntityMeta {
  id: string
  name: string
  description: string
  layout: LayoutType
  fields: FieldMeta[]
  actions: ActionMeta[]
  allow_attachments?: boolean
  version: number
}

export interface EntityRecord {
  id: string
  entity_id: string
  data: Record<string, unknown>
  created_at?: string
  updated_at?: string
}
